nixos/modules/image/amazon.nix

```diff
@@ -0,0 +1,121 @@
+{
+  pkgs,
+  config,
+  lib,
+  ...
+}:
+let
+  cfg = config.image.amazon;
+in
+{
+  imports = [ ./repart.nix ];
+  options = {
+    image.amazon = {
+      bootMode = lib.mkOption {
+        type = lib.types.enum [
+          "uefi-preferred"
+          "uefi"
+        ];
+        default = if pkgs.hostPlatform.isAarch64 then "uefi" else "uefi-preferred";
+      };
+    };
+
+  };
+  config = {
+    boot.loader.grub = {
+      enable = true;
+      efiSupport = true;
+      efiInstallAsRemovable = true;
+      device = if cfg.bootMode == "uefi-preferred" then "/dev/vda" else "nodev";
+    };
+    fileSystems."/" = {
+      fsType = "ext4";
+      device = "/dev/disk/by-partlabel/root";
+    };
+    # Grows on first boot
+    systemd.repart.partitions = {
+      "10-root".repartConfig.Type = "root";
+    };
+    systemd.services.nix-store-load-db = {
+      wantedBy = [ "multi-user.target" ];
+      before = [ "nix-daemon.socket" ];
+      unitConfig.ConditionFirstBoot = true;
+      serviceConfig.ExecStart = "nix-store --load-db < /nix/store/.registration";
+    };
+    system.image.id = config.system.name;
+    system.image.version = config.system.nixos.version;
+    image.repart = {
+      name = config.system.name;
+      partitions = {
+        "00-esp" = {
+          repartConfig = {
+            Type = "esp";
+            Format = "vfat";
+            SizeMinBytes = "1G";
+            SizeMaxBytes = "1G";
+          };
+        };
+        "05-bios" = lib.mkIf (config.image.amazon.bootMode == "uefi-preferred") {
+          repartConfig = {
+            Type = "21686148-6449-6e6f-744e-656564454649";
+            SizeMinBytes = "1M";
+            SizeMaxBytes = "1M";
+          };
+        };
+        "10-root" = {
+          storePaths = [ config.system.build.toplevel ];
+          contents = {
+            # This is needed for nixos-enter to work
+            "/etc/NIXOS".source = pkgs.writeText "NIXOS" "";
+            "/nix/var/nix/profiles".source = pkgs.runCommand "profiles" { } ''
```

Comment:
Not sure if this works. This probably needs to be a proper directory and not a symlink to something in the nix store.